uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/API/Query/LeaveBalancesSelect.php

```diff
@@ -0,0 +1,293 @@
+   *
+   * @param array $params
+   * @param CRM_HRLeaveAndAbsences_Service_LeaveManager $leaveManagerService
+   */
+  public function __construct($params, LeaveManagerService $leaveManagerService) {
+    $this->params = $params;
+    $this->leaveManagerService = $leaveManagerService;
+    $this->buildCustomQuery();
+  }
+
+  /**
+   * Builds the custom query (add joins, where clauses, set params etc)
+   */
+  private function buildCustomQuery() {
+    $customQuery = CRM_Utils_SQL_Select::from(Contact::getTableName() . ' as a');
+
+    $this->addJoins($customQuery);
+    $this->addWhere($customQuery);
+    $this->setReturnFields();
+
+    $this->query = new SelectQuery('Contact', $this->params);
+    $this->query->merge($customQuery);
+  }
+
+  /**
+   * Adds all the JOINs necessary for this query.
+   *
+   * @param CRM_Utils_SQL_Select $query
+   */
+  private function addJoins(CRM_Utils_SQL_Select $query) {
+    $absencePeriod = $this->getAbsencePeriod();
+
+    $joins = [
+      'INNER JOIN ' . HRJobContract::getTableName() . ' jc ON a.id = jc.contact_id',
+      'INNER JOIN ' . HRJobContractRevision::getTableName() . ' jcr ON jcr.id = (SELECT id
+                    FROM ' . HRJobContractRevision::getTableName() . " jcr2
+                    WHERE
+                    jcr2.jobcontract_id = jc.id AND (
+                      jcr2.effective_date <= '{$absencePeriod->start_date}'
+                      OR
+                      (
+                        jcr2.effective_date >= '{$absencePeriod->start_date}' AND
+                        jcr2.effective_date <= '{$absencePeriod->end_date}' 
+                      )
+                    )
+                    ORDER BY jcr2.effective_date DESC, jcr2.id DESC
+                    LIMIT 1)",
+      'INNER JOIN ' . HRJobDetails::getTableName() . ' jd ON jd.jobcontract_revision_id = jcr.details_revision_id'
+    ];
+
+    if($this->getManagerID()) {
+      $joins[] = 'LEFT JOIN ' . Relationship::getTableName() . ' r ON r.contact_id_a = a.id';
+      $joins[] = 'LEFT JOIN ' . RelationshipType::getTableName() . ' rt ON rt.id = r.relationship_type_id';
+    }
+
+    $query->join(null, $joins);
+  }
+
+  /**
+   * Add all the where clauses of the query.
+   *
+   * This method works in combination with addJoins() (that is, it references
+   * tables added by the join clauses).
+   *
+   * @param CRM_Utils_SQL_Select $customQuery
+   */
+  private function addWhere(CRM_Utils_SQL_Select $customQuery) {
+    $absencePeriod = $this->getAbsencePeriod();
+
+    $conditions = [
+      'a.is_deleted = 0',
+      'jc.deleted = 0',
+      "(
+        (jd.period_end_date IS NOT NULL AND jd.period_start_date <= '{$absencePeriod->end_date}' AND jd.period_end_date >= '{$absencePeriod->start_date}')
+          OR
+        (jd.period_end_date IS NULL AND 
+          (
+            (jd.period_start_date >= '{$absencePeriod->start_date}' AND jd.period_start_date <= '{$absencePeriod->end_date}')
```

Comment:
Do we really need this condition? Once the jd.period_end_date IS NULL, I'm thinking the only criteria left to be fulfilled is that the jd.period_start_date is less than the Absence period end date.